Update Go to v1.20

### What happened?

When trying to update a provider that uses this library, I ran into an error:
```
../../../../go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/pf@v0.9.2/internal/muxer/muxer.go:75:33: undefined: errors.Join
```

This is because https://github.com/pulumi/pulumi-terraform-bridge/blob/902e9d5f3c181c4a64079d06e2ae9adf0d474696/pf/internal/muxer/muxer.go#L75 uses the new `errors.Join` function which was introduced in Go v1.20. Any downstream projects that uses this function without compiling with Go v1.20 will error out.

We should update our go.mod files to indicate that v1.20 of Go should be used, since this breaks backwards compatibility with Go 1.19 and below.

### Expected Behavior

Build/compilation to be successful, or a warning message indicating that the Golang compiler is a different version to the one declared by this project's go.mod.

### Steps to reproduce

Use `tfgen.MainWithMuxer` on Golang v1.19.X or below. Eg. `make tfgen` in the pulumi-gcp providers repo.

### Output of `pulumi about`

-

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -10,7 +10,12 @@ jobs:
     name: Build and Test Bridge
     strategy:
       matrix:
-        go-version: [1.20.x]
+        # To avoid depending on features introduced in newer golang versions, we need to
+        # test our minimum supported golang versions.
+        #
+        # When we decide to bump our minimum go version, we need to remember to bump the
+        # go version in our go.mod files.
+        go-version: [1.19.x, 1.20.x]
         platform: [ubuntu-latest, macos-latest, windows-latest]
     runs-on: ${{ matrix.platform }}
     steps:

--- a/pf/internal/muxer/muxer.go
+++ b/pf/internal/muxer/muxer.go
@@ -15,7 +15,6 @@
 package muxer
 
 import (
-	"errors"
 	"fmt"
 	"sort"
 	"strings"
@@ -72,7 +71,14 @@ func AugmentShimWithDisjointPF(ctx context.Context, shim shim.Provider, pf provi
 	if len(resources) > 0 {
 		rErr = fmt.Errorf("ResourcesMap is not disjoint: conflicting keys: %s", strings.Join(resources, ", "))
 	}
-	contract.AssertNoErrorf(errors.Join(rErr, dErr), "providers are not disjoint")
+	switch {
+	case dErr != nil && rErr != nil:
+		contract.Failf("Providers are not disjoint:\n- %s\n- %s", rErr.Error(), dErr.Error())
+	case rErr != nil:
+		contract.Failf("Resources are not disjoint: %s", rErr.Error())
+	case dErr != nil:
+		contract.Failf("Datasources are not disjoint: %s", dErr.Error())
+	}
 
 	return p
 }

--- a/pf/tests/main_test.go
+++ b/pf/tests/main_test.go
@@ -15,7 +15,6 @@
 package tfbridgetests
 
 import (
-	"errors"
 	"fmt"
 	"log"
 	"os"
@@ -43,7 +42,15 @@ func testMain(m *testing.M) (exitCode int, err error) {
 			panicError = fmt.Errorf("Ignoring panic in tests: %v", panicResult)
 		}
 		teardownError := teardownUnitTests()
-		err = errors.Join(panicError, teardownError)
+		if teardownError == nil {
+			err = panicError
+		} else {
+			// Wrapping multiple errors was introduced on v1.20.0.  While we
+			// still support go v1.19.0, we need to convert one error into a
+			// string.
+			err = fmt.Errorf("Tests panicked and teardown failed: %w; %s",
+				panicError, teardownError.Error())
+		}
 	}()
 
 	exitCode = m.Run()

--- a/pf/tfbridge/ignore_changes.go
+++ b/pf/tfbridge/ignore_changes.go
@@ -15,9 +15,9 @@
 package tfbridge
 
 import (
-	"errors"
 	"fmt"
 
+	"github.com/hashicorp/go-multierror"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
 )
 
@@ -33,8 +33,8 @@ func applyIgnoreChanges(old, new resource.PropertyMap, ignoreChanges []string) (
 		}
 		paths = append(paths, pp)
 	}
-	if err := errors.Join(errs...); err != nil {
-		return nil, err
+	if len(errs) > 0 {
+		return nil, &multierror.Error{Errors: errs}
 	}
 
 	newValue := resource.NewObjectProperty(new.Copy())

--- a/x/muxer/muxer.go
+++ b/x/muxer/muxer.go
@@ -452,7 +452,8 @@ func (m *muxer) GetMapping(ctx context.Context, req *rpc.GetMappingRequest) (*rp
 	result, err := combineMapping(&args)
 	if err != nil {
 		if args.err != nil {
-			return nil, fmt.Errorf("%w (sub-provider GetMapping call failed: %w)", err, args.err)
+			// go v1.19.0 only accepts a single %w within fmt.Error, so we convert the second call to %s.
+			return nil, fmt.Errorf("%w (sub-provider GetMapping call failed: %s)", err, args.err.Error())
 		}
 		return nil, err
 	}

